rate/rate.go

```diff
@@ -0,0 +1,105 @@
+package rate
+
+import (
+	"sync"
+	"time"
+
+	"github.com/newrelic/newrelic-telemetry-sdk-go/internal"
+	"github.com/newrelic/newrelic-telemetry-sdk-go/telemetry"
+)
+
+// RateCalculator is used to create Gauge metrics from rate values.
+type RateCalculator struct {
+	lock                    sync.Mutex
+	datapoints              internal.Datapoints
+	lastClean               time.Time
+	expirationCheckInterval time.Duration
+	expirationAge           time.Duration
+}
+
+// NewRateCalculator creates a new RateCalculator.  A single RateCalculator
+// stores the last timestamp in order to compute the rate.
+func NewRateCalculator() *RateCalculator {
+	return &RateCalculator{
+		datapoints: internal.Datapoints{},
+		// These defaults are described in the Set method doc comments.
+		expirationCheckInterval: 20 * time.Minute,
+		expirationAge:           20 * time.Minute,
+	}
+}
+
+// GetRate creates a Gauge metric with rate of change based on the previous timestamp.
+// If no previous timestamp is NOT found, returns false (as no calculation is made)
+// If a previous timestamp is found use it to get the elapsed time (in seconds) and use that as the denominator
+// Rate = value / (now - before)[s]
+func (rc *RateCalculator) GetRate(name string, attributes map[string]interface{}, val float64, now time.Time) (gauge telemetry.Gauge, valid bool) {
```

Comment:
Exporting a rate from a remote location only serves to decrease the usability of this data. You cannot reverse this operation without also transmitting the time interval it was calculated with. New Relic has the functionality to take one of the existing Metrics and calculate rates (as well as other computed values) from these.

Additionally, this would add an inconsistency in this `telemetry-sdk`s API from the rest of the New Relic `telemetry-sdk`s. If we want to add this metric type it need to be addressed in [the specification](https://github.com/newrelic/newrelic-telemetry-sdk-specs)